### PR TITLE
feat: deep typeguard about destructor

### PR DESCRIPTION
**Pull request by sunrabbit123**
**Description:**
```ts
//@strict: true

type Func = <T extends ["a", number] | ["b", string]>(...args: T) => void;

const f60: Func = (kind, payload) => {
  if (kind === "a") {
    payload.toFixed(); // error
  }
  if (kind === "b") {
    payload.toUpperCase(); // error
  }
};
```

### CODE REVIEW

File: crates/stc_ts_file_analyzer/src/analyzer/expr/function.rs [hunk 1]
```diff
@@ -179,16 +179,56 @@ impl Analyzer<'_, '_> {
                         });
                     }
                 }
+                match (*param.ty).normalize() {
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/expr/function.rs:182]:
```suggestion
                match param.ty.normalize() {
```

File: crates/stc_ts_file_analyzer/src/analyzer/expr/function.rs [hunk 2]
```diff
@@ -179,16 +179,56 @@ impl Analyzer<'_, '_> {
...
                 tracker: Default::default(),
             });
+            params_tuple.freeze();
+
+            let destructure_key = self.get_destructor_unique_key();
 
             for (idx, param) in params.enumerate() {
+                if temp_els.len() == 1 {
+                    if let Some(TupleElement { ty: box ty, .. }) = temp_els.first_mut() {
+                        if let Type::Union(Union { types, .. }) = ty.normalize_mut() {
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/expr/function.rs:220]: Please use `if let Some(Union { types, .. }) = ty.as_union_type_mut()` instead

File: crates/stc_ts_file_analyzer/src/analyzer/expr/function.rs [hunk 3]
```diff
@@ -179,16 +179,56 @@ impl Analyzer<'_, '_> {
...
+                            for ty in types {
+                                if let Type::Tuple(Tuple { elems, .. }) = ty.normalize_mut() {
```
Comment (kdy1) [inline crates/stc_ts_file_analyzer/src/analyzer/expr/function.rs:222]: Please use if let Some(Tuple { elems, .. }) = ty.as_union_type_mut() instead

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thank you!

**Comment by sunrabbit123** [comment]
fixed it

**Review by kdy1 — DISMISSED** [review]
Thanks!

**Review by kdy1 — APPROVED** [review]

**Pull request merged**